File: packages/hermes-plugin/src/dmemo_hermes/journal.py
"""Journaling VectorStore wrapper (D7) — Python port of ``store/journal.ts``.

Wraps mem0's native Python ``VectorStoreBase`` instance and:

 1. forwards every call through unchanged, so search/read behaviour is
    identical to un-wrapped mem0;
 2. records every mutating call as a spec ``VectorOp`` for ``flush()`` to
    drain;
 3. keeps a materialized ``{id: (vector, payload)}`` mirror, because mem0's
    stores do not return embedding vectors from ``get()``/``list()``
    (gotcha 16 — FAISS's ``OutputData`` is ``id``/``score``/``payload``
    only), and checkpoint blobs need full rows.

Installed by post-init property swap (``memory.vector_store = Journaling...``),
the same shape as the TypeScript adapter: mem0 has no custom-store
registration hook in either language (gotcha 11).
"""

from __future__ import annotations

import datetime as _dt
from typing import Any, Dict, List, Optional, Sequence, Tuple

from .blob import pack_vector, unpack_vector


def _utc_now_iso() -> str:
    return _dt.datetime.now(_dt.timezone.utc).isoformat().replace("+00:00", "Z")


class JournalingVectorStore:
    """Mutation-journaling proxy around a mem0 ``VectorStoreBase``."""

    def __init__(self, native: Any) -> None:
        self._native = native
        self._journal: List[Dict[str, Any]] = []
        self._mirror: Dict[str, Tuple[List[float], Dict[str, Any]]] = {}

    # -- reads / lifecycle: delegate untouched, never journaled -------------

    def search(self, query: str, vectors: Sequence[float], top_k: int = 5, filters: Optional[Dict] = None):
        return self._native.search(query, vectors, top_k, filters)

    def keyword_search(self, query: str, top_k: int = 5, filters: Optional[Dict] = None):
        return self._native.keyword_search(query, top_k, filters)

    def get(self, vector_id: str):
        return self._native.get(vector_id)

    def list(self, filters: Optional[Dict] = None, top_k: Optional[int] = None):
        return self._native.list(filters, top_k)

    def list_cols(self):
        return self._native.list_cols()

    def col_info(self):
        return self._native.col_info()

    def create_col(self, *args: Any, **kwargs: Any):
        # Collection creation is a store-local concern; the blob chain
        # describes rows, not collections.
        return self._native.create_col(*args, **kwargs)

    def __getattr__(self, name: str) -> Any:
        # Store-specific extras (``_save``, ``client``, ``index`` …) that mem0
        # or a host may reach for. Only consulted for attributes this class
        # does not define, so it can never shadow the journaled methods.
        return getattr(self._native, name)

    # -- mutations: forward to native, then journal + mirror ----------------

    def insert(self, vectors: Sequence[Sequence[float]], payloads: Optional[Sequence[Dict]] = None, ids: Optional[Sequence[str]] = None):
        result = self._native.insert(vectors, payloads, ids)
        ids = list(ids or [])
        payloads = [dict(p) for p in (payloads or [{} for _ in vectors])]
        for i, vector_id in enumerate(ids):
            self._mirror[vector_id] = (list(vectors[i]), payloads[i])
        self._journal.append(
            {
                "op": "insert",
                "ids": ids,
                "vectors": [pack_vector(v) for v in vectors],
                "payloads": payloads,
            }
        )
        return result

    def update(self, vector_id: str, vector: Optional[Sequence[float]] = None, payload: Optional[Dict] = None):
        result = self._native.update(vector_id, vector, payload)
        existing = self._mirror.get(vector_id)
        new_vector = list(vector) if vector is not None else (list(existing[0]) if existing else [])
        new_payload = dict(payload) if payload is not None else (dict(existing[1]) if existing else {})
        self._mirror[vector_id] = (new_vector, new_payload)
        # The spec's update op always carries a vector, so a payload-only
        # mem0 update still replays correctly on a host that never saw the
        # original insert.
        self._journal.append(
            {"op": "update", "id": vector_id, "vector": pack_vector(new_vector), "payload": new_payload}
        )
        return result

    def delete(self, vector_id: str):
        result = self._native.delete(vector_id)
        self._mirror.pop(vector_id, None)
        self._journal.append({"op": "delete", "id": vector_id})
        return result

    def delete_col(self):
        result = self._native.delete_col()
        self._mirror.clear()
        self._journal.append({"op": "deleteCol"})
        return result

    def reset(self):
        # mem0's ``reset()`` is delete-collection-then-recreate; on the wire
        # that is exactly a deleteCol op — recreation carries no rows.
        result = self._native.reset()
        self._mirror.clear()
        self._journal.append({"op": "deleteCol"})
        return result

    # -- dmemo-specific extensions -----------------------------------------

    def journal_tombstone(self, epoch: int, reason: Optional[str] = None) -> None:
        op: Dict[str, Any] = {"op": "tombstone", "epoch": epoch, "tombstonedAt": _utc_now_iso()}
        if reason is not None:
            op["reason"] = reason
        self._journal.append(op)

    def drain_journal(self) -> List[Dict[str, Any]]:
        drained, self._journal = self._journal, []
        return drained

    def has_pending_ops(self) -> bool:
        return len(self._journal) > 0

    def snapshot_rows(self) -> List[Dict[str, Any]]:
        """Full materialized vector rows, for a checkpoint blob."""
        return [
            {"id": vector_id, "vector": pack_vector(vector), "payload": payload}
            for vector_id, (vector, payload) in self._mirror.items()
        ]

    @property
    def row_count(self) -> int:
        return len(self._mirror)

    # -- restore-time replay (must NOT re-journal) --------------------------

    def apply_replay_op(self, op: Dict[str, Any]) -> None:
        kind = op["op"]
        if kind == "insert":
            vectors = [unpack_vector(v) for v in op["vectors"]]
            payloads = [dict(p) for p in op["payloads"]]
            ids = list(op["ids"])
            self._native.insert(vectors, payloads, ids)
            for i, vector_id in enumerate(ids):
                self._mirror[vector_id] = (vectors[i], payloads[i])
        elif kind == "update":
            vector = unpack_vector(op["vector"])
            payload = dict(op["payload"])
            if op["id"] in self._mirror:
                self._native.update(op["id"], vector, payload)
            else:
                # A chain can legitimately update a row this store never saw
                # (e.g. the insert landed in a blob that a re-embed migration
                # replaced). mem0's stores raise on unknown ids, so replay as
                # an insert rather than aborting the whole chain.
                self._native.insert([vector], [payload], [op["id"]])
            self._mirror[op["id"]] = (vector, payload)
        elif kind == "delete":
            if op["id"] in self._mirror:
                self._native.delete(op["id"])
            self._mirror.pop(op["id"], None)
        elif kind == "deleteCol":
            self._native.delete_col()
            self._mirror.clear()
        elif kind == "tombstone":
            return  # durability handled by preserving the op, no store effect
        else:
            raise ValueError(f"unknown VectorOp: {kind!r}")

    def apply_checkpoint_rows(self, rows: Sequence[Dict[str, Any]]) -> None:
        if not rows:
            return
        ids = [r["id"] for r in rows]
        vectors = [unpack_vector(r["vector"]) for r in rows]
        payloads = [dict(r["payload"]) for r in rows]
        self._native.insert(vectors, payloads, ids)
        for i, vector_id in enumerate(ids):
            self._mirror[vector_id] = (vectors[i], payloads[i])

File: packages/core/src/store/journal.ts
import type { VectorStore, VectorStoreResult, SearchFilters } from 'mem0ai/oss';
import {
  packVector,
  unpackVector,
  type VectorOp,
  type CheckpointVectorRow,
} from '@dmemo/blob-spec';

/**
 * T1.3 — journaling VectorStore wrapper (D7).
 *
 * Wraps mem0's native `VectorStore` instance and:
 *  1. Forwards every call to the native store unchanged (so search/read
 *     behavior is byte-for-byte identical to un-wrapped mem0).
 *  2. Records every mutating call (`insert`/`update`/`delete`/`deleteCol`)
 *     as a spec `VectorOp` (blob-spec's `packVector` for the base64
 *     Float32Array encoding, gotcha 14) into an in-memory journal that
 *     `flush()` drains.
 *  3. Maintains a materialized `Map<id, {vector, payload}>` mirror of
 *     current state. This exists because mem0ai@3.1.1's native store
 *     `list()`/`get()` do not return the embedding vector (only
 *     `{id, payload}`) — there is no public API to read back full vector
 *     rows, which checkpoint blobs need. The mirror is updated on every
 *     intercepted write and is the sole source for `snapshotRows()`.
 *
 * Installed via post-init property swap (`memory.vectorStore = new
 * JournalingVectorStore(nativeStore)`), matching the pattern proven live in
 * `spike/c3-mem0-loop.mjs`: mem0ai 3.1.1's `VectorStoreFactory` has no
 * custom-provider registration path (gotcha 11), so there's no cleaner
 * injection point in this SDK version.
 */
export class JournalingVectorStore implements VectorStore {
  private readonly native: VectorStore;
  private journal: VectorOp[] = [];
  private readonly mirror = new Map<string, { vector: number[]; payload: Record<string, unknown> }>();

  constructor(native: VectorStore) {
    this.native = native;
  }

  // -- lifecycle / reads: delegate untouched, never journaled -------------
  async initialize(): Promise<void> {
    return this.native.initialize();
  }
  async search(query: number[], topK?: number, filters?: SearchFilters): Promise<VectorStoreResult[]> {
    return this.native.search(query, topK, filters);
  }
  async keywordSearch(query: string, topK?: number, filters?: SearchFilters): Promise<VectorStoreResult[] | null> {
    if (!this.native.keywordSearch) return null;
    return this.native.keywordSearch(query, topK, filters);
  }
  async get(vectorId: string): Promise<VectorStoreResult | null> {
    return this.native.get(vectorId);
  }
  async list(filters?: SearchFilters, topK?: number): Promise<[VectorStoreResult[], number]> {
    return this.native.list(filters, topK);
  }
  async getUserId(): Promise<string> {
    return this.native.getUserId();
  }
  async setUserId(userId: string): Promise<void> {
    return this.native.setUserId(userId);
  }

  // -- mutations: forward to native, then journal + mirror -----------------
  async insert(vectors: number[][], ids: string[], payloads: Record<string, unknown>[]): Promise<void> {
    await this.native.insert(vectors, ids, payloads);
    for (let i = 0; i < ids.length; i++) {
      const id = ids[i]!;
      const vector = vectors[i]!;
      const payload = payloads[i] ?? {};
      this.mirror.set(id, { vector, payload });
    }
    this.journal.push({
      op: 'insert',
      ids: [...ids],
      vectors: vectors.map(packVector),
      payloads: payloads.map((p) => ({ ...p })),
    });
  }

  async update(vectorId: string, vector: number[], payload: Record<string, unknown>): Promise<void> {
    await this.native.update(vectorId, vector, payload);
    this.mirror.set(vectorId, { vector, payload });
    this.journal.push({ op: 'update', id: vectorId, vector: packVector(vector), payload: { ...payload } });
  }

  async delete(vectorId: string): Promise<void> {
    await this.native.delete(vectorId);
    this.mirror.delete(vectorId);
    this.journal.push({ op: 'delete', id: vectorId });
  }

  async deleteCol(): Promise<void> {
    await this.native.deleteCol();
    this.mirror.clear();
    this.journal.push({ op: 'deleteCol' });
  }

  // -- dmemo-specific extensions --------------------------------------------

  /** T1.7: record a forget-epoch tombstone. Carries no vector-store mutation
   * (nothing is deleted from the native store or mirror by this call alone —
   * callers that want to actually purge rows still issue explicit `delete`
   * calls; the tombstone is the durable, on-chain-visible marker). */
  journalTombstone(epoch: number, reason?: string): void {
    const op: VectorOp = {
      op: 'tombstone',
      epoch,
      tombstonedAt: new Date().toISOString(),
      ...(reason !== undefined ? { reason } : {}),
    };
    this.journal.push(op);
  }

  /** Drain and return all journaled ops since the last drain (or since
   * construction). Used by `flush()` to build a delta blob. */
  drainJournal(): VectorOp[] {
    const j = this.journal;
    this.journal = [];
    return j;
  }

  /** True if there are unflushed ops pending (used for flush-cadence checks
   * without mutating the journal). */
  hasPendingOps(): boolean {
    return this.journal.length > 0;
  }

  /** Full materialized vector rows, for building a checkpoint blob. */
  snapshotRows(): CheckpointVectorRow[] {
    const rows: CheckpointVectorRow[] = [];
    for (const [id, { vector, payload }] of this.mirror.entries()) {
      rows.push({ id, vector: packVector(vector), payload });
    }
    return rows;
  }

  get rowCount(): number {
    return this.mirror.size;
  }

  /**
   * Apply a previously-journaled op directly to the native store + mirror,
   * WITHOUT re-journaling it. Used exclusively during restore-time replay
   * (walking the delta/checkpoint chain into a fresh native store) — replay
   * must not itself produce new journal entries.
   */
  async applyReplayOp(op: VectorOp): Promise<void> {
    switch (op.op) {
      case 'insert': {
        const vectors = op.vectors.map(unpackVector);
        await this.native.insert(vectors, op.ids, op.payloads);
        for (let i = 0; i < op.ids.length; i++) {
          const id = op.ids[i]!;
          this.mirror.set(id, { vector: vectors[i]!, payload: op.payloads[i] ?? {} });
        }
        return;
      }
      case 'update': {
        const vector = unpackVector(op.vector);
        await this.native.update(op.id, vector, op.payload);
        this.mirror.set(op.id, { vector, payload: op.payload });
        return;
      }
      case 'delete': {
        await this.native.delete(op.id);
        this.mirror.delete(op.id);
        return;
      }
      case 'deleteCol': {
        await this.native.deleteCol();
        this.mirror.clear();
        return;
      }
      case 'tombstone':
        // No vector-store side effect on replay — durability is handled by
        // the caller preserving tombstone ops in the replayed journal state.
        return;
      default: {
        const _exhaustive: never = op;
        throw new Error(`unknown VectorOp: ${JSON.stringify(_exhaustive)}`);
      }
    }
  }

  /** Load a full checkpoint's rows directly into a fresh native store +
   * mirror (bypasses per-row insert journaling; used once at the start of
   * chain replay when the chain begins with a checkpoint). */
  async applyCheckpointRows(rows: CheckpointVectorRow[]): Promise<void> {
    if (rows.length === 0) return;
    const ids = rows.map((r) => r.id);
    const vectors = rows.map((r) => unpackVector(r.vector));
    const payloads = rows.map((r) => r.payload);
    await this.native.insert(vectors, ids, payloads);
    for (let i = 0; i < ids.length; i++) {
      this.mirror.set(ids[i]!, { vector: vectors[i]!, payload: payloads[i]! });
    }
  }
}
